devicetypes/smartthings/zwave-radiator-thermostat.src/zwave-radiator-thermostat.groovy

```diff
@@ -0,0 +1,262 @@
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.sensormultilevelv5.SensorMultilevelReport cmd) {
+	def map = [name: "temperature", value: convertTemperatureIfNeeded(cmd.scaledSensorValue, 'C', cmd.precision), unit: temperatureScale]
+	createEvent(map)
+}
+
+def zwaveEvent(physicalgraph.zwave.Command cmd) {
+	log.warn "Unhandled command: ${cmd}"
+	[:]
+}
+
+def setThermostatMode(String mode) {
+	def modeValue = 0
+	switch (mode) {
+		case "heat":
+			modeValue = 1
+			break
+		case "emergency heat":
+			modeValue = 11
+			break
+		case "off":
+			modeValue = 0
+			break
+	}
+
+	[
+			secure(zwave.thermostatModeV2.thermostatModeSet(mode: modeValue)),
+			"delay 2000",
+			secure(zwave.thermostatModeV2.thermostatModeGet())
+	]
+}
+
+def heat() {
+	setThermostatMode("heat")
+}
+
+def emergencyHeat() {
+	setThermostatMode("emergency heat")
+}
+
+def off() {
+	setThermostatMode("off")
+}
+
+def setHeatingSetpoint(setpoint) {
+	setpoint = temperatureScale == 'C' ? setpoint : fahrenheitToCelsius(setpoint)
+	[
+			secure(zwave.thermostatSetpointV2.thermostatSetpointSet([precision: 1, scale: 0, scaledValue: setpoint, setpointType: 1, size: 2])),
+			"delay 2000",
+			secure(zwave.thermostatSetpointV2.thermostatSetpointGet(setpointType: 1))
+	]
+}
+
+def setThermostatSetpointUp() {
+	def setpoint = device.latestValue("heatingSetpoint")
+	if (setpoint < maxHeatingSetpointTemperature) {
+		setpoint = setpoint + (temperatureScale == 'C' ? 0.5 : 1)
+	}
+	setHeatingSetpoint(setpoint)
+}
+
+def setThermostatSetpointDown() {
+	def setpoint = device.latestValue("heatingSetpoint")
+	if (setpoint > minHeatingSetpointTemperature) {
+		setpoint = setpoint - (temperatureScale == 'C' ? 0.5 : 1)
+	}
+	setHeatingSetpoint(setpoint)
+}
+
+def refresh() {
+	def cmds = [
+			secure(zwave.batteryV1.batteryGet()),
+			secure(zwave.thermostatSetpointV2.thermostatSetpointGet(setpointType: 1)),
+			secure(zwave.sensorMultilevelV5.sensorMultilevelGet()),
+			secure(zwave.thermostatModeV2.thermostatModeGet())
+	]
+
+	delayBetween(cmds, 2500)
+}
+
+def ping() {
+	refresh()
+}
+
+private secure(cmd) {
+	if (zwaveInfo.zw.endsWith("s")) {
+		zwave.securityV1.securityMessageEncapsulation().encapsulate(cmd).format()
+	} else {
+		cmd.format()
+	}
+}
+
+def switchMode() {
+	def currentMode = device.currentValue("thermostatMode")
+	def supportedModes = thermostatSupportedModes
+	if (supportedModes && supportedModes.size()) {
+		def next = { supportedModes[supportedModes.indexOf(it) + 1] ?: supportedModes[0] }
+		def nextMode = next(currentMode)
+		setThermostatMode(nextMode)
+	} else {
+		log.warn "supportedModes not defined"
+	}
+}
+
+private getMaxHeatingSetpointTemperature() {
+	temperatureScale == 'C' ? 30 : 86
+}
+
+private getMinHeatingSetpointTemperature() {
+	temperatureScale == 'C' ? 10 : 50
+}
+
+private getThermostatSupportedModes() {
+	if (isEverspringRadiatorThermostat()) {
+		["off", "heat", "emergency heat"]
+	} else {
+		[]
```

Comment:
We should probably have some sort of default here, even if there are no other fingerprints in the DTH, since this DTH is designed to be generic. I'd just have the else return `["off","heat"]`.